File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/state/solution/square.cpp
// square.cpp

#include <iostream>
#include "dp_exception.h"
#include "game.h"
#include "square.h"

square::square(bool a_has_a_mine, int a_row, int a_column)
: my_has_a_mine(a_has_a_mine)
, my_row(a_row)
, my_column(a_column)
, my_state(new state_1_covered())
{
}

bool square::is_covered() const
{
	return my_state->is_covered();
}

bool square::has_a_mine() const
{
	return my_has_a_mine;
}

void square::left_click(game & a_game)
{
	my_state->left_click(a_game, *this);
}

void square::right_click()
{
	my_state->right_click(*this);
}

void square::output(const game & a_game) const
{
	my_state->output(a_game, *this);
}

void square::uncover()
{
	my_state->uncover(*this);
}

////////////////////////////////////////////////////////////////////////////////

bool square::state_1_covered::is_covered() const
{
	return true;
}

void square::state_1_covered::left_click(game & a_game, square & a_square) const
{
	if (a_square.my_has_a_mine)
	{
		a_square.my_state = new state_5_uncovered_with_mine();
		a_game.uncover_a_mine();
	}
	else
	{
		a_square.my_state = new state_4_uncovered_without_mine();
		a_game.uncover_a_non_mine();
	}
}

void square::state_1_covered::right_click(square & a_square) const
{
	a_square.my_state = new state_2_covered_with_flag();
}

void square::state_1_covered::output(const game & a_game, const square & a_square) const
{
	std::cout << 'C';
}

void square::state_1_covered::uncover(square & a_square) const
{
	if (a_square.my_has_a_mine)
	{
		a_square.my_state = new state_5_uncovered_with_mine();
	}
	else
	{
		a_square.my_state = new state_4_uncovered_without_mine();
	}
}

////////////////////////////////////////////////////////////////////////////////

bool square::state_2_covered_with_flag::is_covered() const
{
	return true;
}

void square::state_2_covered_with_flag::left_click(game & a_game, square & a_square) const
{
}

void square::state_2_covered_with_flag::right_click(square & a_square) const
{
	a_square.my_state = new state_3_covered_with_question_mark();
}

void square::state_2_covered_with_flag::output(const game & a_game, const square & a_square) const
{
	std::cout << 'F';
}

void square::state_2_covered_with_flag::uncover(square & a_square) const
{
	if (a_square.my_has_a_mine)
	{
		a_square.my_state = new state_5_uncovered_with_mine();
	}
	else
	{
		a_square.my_state = new state_4_uncovered_without_mine();
	}
}

////////////////////////////////////////////////////////////////////////////////

bool square::state_3_covered_with_question_mark::is_covered() const
{
	return true;
}

void square::state_3_covered_with_question_mark::left_click(game & a_game, square & a_square) const
{
	if (a_square.my_has_a_mine)
	{
		a_square.my_state = new state_5_uncovered_with_mine();
		a_game.uncover_a_mine();
	}
	else
	{
		a_square.my_state = new state_4_uncovered_without_mine();
		a_game.uncover_a_non_mine();
	}
}

void square::state_3_covered_with_question_mark::right_click(square & a_square) const
{
	a_square.my_state = new state_1_covered();
}

void square::state_3_covered_with_question_mark::output(const game & a_game, const square & a_square) const
{
	std::cout << '?';
}

void square::state_3_covered_with_question_mark::uncover(square & a_square) const
{
	if (a_square.my_has_a_mine)
	{
		a_square.my_state = new state_5_uncovered_with_mine();
	}
	else
	{
		a_square.my_state = new state_4_uncovered_without_mine();
	}
}

////////////////////////////////////////////////////////////////////////////////

bool square::state_4_uncovered_without_mine::is_covered() const
{
	return false;
}

void square::state_4_uncovered_without_mine::left_click(game & a_game, square & a_square) const
{
}

void square::state_4_uncovered_without_mine::right_click(square & a_square) const
{
}

void square::state_4_uncovered_without_mine::output(const game & a_game, const square & a_square) const
{
	int n = a_game.get_number_of_neighbors_who_have_a_mine(a_square.my_row, a_square.my_column);
	if (n == 0)
	{
		std::cout << ' ';
	}
	else
	{
		std::cout << char('0' + n);
	}
}

void square::state_4_uncovered_without_mine::uncover(square & a_square) const
{
}

////////////////////////////////////////////////////////////////////////////////

bool square::state_5_uncovered_with_mine::is_covered() const
{
	return false;
}

void square::state_5_uncovered_with_mine::left_click(game & a_game, square & a_square) const
{
}

void square::state_5_uncovered_with_mine::right_click(square & a_square) const
{
}

void square::state_5_uncovered_with_mine::output(const game & a_game, const square & a_square) const
{
	std::cout << 'M';
}

void square::state_5_uncovered_with_mine::uncover(square & a_square) const
{
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/state/solution/Square.java
// Square.java

public class Square
{
	public Square(boolean aHasAMine, int aRow, int aColumn)
	{
		myState = new State1Covered();
		myHasAMine = aHasAMine;
		myRow = aRow;
		myColumn = aColumn;
	}

	public boolean isCovered()
	{
		return myState.isCovered();
	}

	public boolean hasAMine()
	{
		return myHasAMine;
	}

	public void leftClick(Game aGame)
	{
		myState.leftClick(aGame, this);
	}

	public void rightClick()
	{
		myState.rightClick(this);
	}

	public void output(Game aGame)
	{
		myState.output(aGame, this);
	}

	public void uncover()
	{
		myState.uncover(this);
	}

	private State myState;
	private boolean myHasAMine;
	private int myRow;
	private int myColumn;

	////////////////////////////////////////////////////////////////////////////////

	private abstract static class State
	{
		public abstract boolean isCovered();
		public abstract void leftClick(Game aGame, Square aSquare);
		public abstract void rightClick(Square aSquare);
		public abstract void output(Game aGame, Square aSquare);
		public abstract void uncover(Square aSquare);
	}

	////////////////////////////////////////////////////////////////////////////////

	private static class State1Covered extends State
	{
		public boolean isCovered()
		{
			return true;
		}

		public void leftClick(Game aGame, Square aSquare)
		{
			if (aSquare.myHasAMine)
			{
				aSquare.myState = new State5UncoveredWithMine();
				aGame.uncoverAMine();
			}
			else
			{
				aSquare.myState = new State4UncoveredWithoutMine();
				aGame.uncoverANonMine();
			}
		}

		public void rightClick(Square aSquare)
		{
			aSquare.myState = new State2CoveredWithFlag();
		}

		public void output(Game aGame, Square aSquare)
		{
			System.out.print('C');
		}

		public void uncover(Square aSquare)
		{
			if (aSquare.myHasAMine)
			{
				aSquare.myState = new State5UncoveredWithMine();
			}
			else
			{
				aSquare.myState = new State4UncoveredWithoutMine();
			}
		}
	}

	////////////////////////////////////////////////////////////////////////////////

	private static class State2CoveredWithFlag extends State
	{
		public boolean isCovered()
		{
			return true;
		}

		public void leftClick(Game aGame, Square aSquare)
		{
		}

		public void rightClick(Square aSquare)
		{
			aSquare.myState = new State3CoveredWithQuestionMark();
		}

		public void output(Game aGame, Square aSquare)
		{
			System.out.print('F');
		}

		public void uncover(Square aSquare)
		{
			if (aSquare.myHasAMine)
			{
				aSquare.myState = new State5UncoveredWithMine();
			}
			else
			{
				aSquare.myState = new State4UncoveredWithoutMine();
			}
		}
	}

	////////////////////////////////////////////////////////////////////////////////

	private static class State3CoveredWithQuestionMark extends State
	{
		public boolean isCovered()
		{
			return true;
		}

		public void leftClick(Game aGame, Square aSquare)
		{
			if (aSquare.myHasAMine)
			{
				aSquare.myState = new State5UncoveredWithMine();
				aGame.uncoverAMine();
			}
			else
			{
				aSquare.myState = new State4UncoveredWithoutMine();
				aGame.uncoverANonMine();
			}
		}

		public void rightClick(Square aSquare)
		{
			aSquare.myState = new State1Covered();
		}

		public void output(Game aGame, Square aSquare)
		{
			System.out.print('?');
		}

		public void uncover(Square aSquare)
		{
			if (aSquare.myHasAMine)
			{
				aSquare.myState = new State5UncoveredWithMine();
			}
			else
			{
				aSquare.myState = new State4UncoveredWithoutMine();
			}
		}
	}

	////////////////////////////////////////////////////////////////////////////////

	private static class State4UncoveredWithoutMine extends State
	{
		public boolean isCovered()
		{
			return false;
		}

		public void leftClick(Game aGame, Square aSquare)
		{
		}

		public void rightClick(Square aSquare)
		{
		}

		public void output(Game aGame, Square aSquare)
		{
			int n = aGame.getNumberOfNeighborsWhoHaveAMine(aSquare.myRow, aSquare.myColumn);
			if (n == 0)
			{
				System.out.print(' ');
			}
			else
			{
				System.out.print((char) ('0' + n));
			}
		}

		public void uncover(Square aSquare)
		{
		}
	}

	////////////////////////////////////////////////////////////////////////////////

	private static class State5UncoveredWithMine extends State
	{
		public boolean isCovered()
		{
			return false;
		}

		public void leftClick(Game aGame, Square aSquare)
		{
		}

		public void rightClick(Square aSquare)
		{
		}

		public void output(Game aGame, Square aSquare)
		{
			System.out.print('M');
		}

		public void uncover(Square aSquare)
		{
		}
	}
}
